Firestore/core/test/unit/bundle/bundle_reader_test.cc

```diff
@@ -0,0 +1,533 @@
+using ProtoMaybeDocument = ::firestore::client::MaybeDocument;
+using ProtoNamedQuery = ::firestore::NamedQuery;
+using ProtoValue = ::google::firestore::v1::Value;
+using model::DatabaseId;
+using nanopb::ProtobufParse;
+
+class BundleReaderTest : public ::testing::Test {
+ public:
+  BundleReaderTest()
+      : remote_serializer(DatabaseId("p", "default")),
+        local_serializer(remote_serializer),
+        bundle_serializer(remote_serializer) {
+    msg_diff_.ReportDifferencesToString(&message_differences);
+  }
+
+  static std::string FullPath(const std::string& path) {
+    return "projects/p/databases/default/documents/" + path;
+  }
+
+  std::string AddNamedQuery(const ProtoNamedQuery& data) {
+    std::string json;
+    ProtoBundleElement element;
+    *element.mutable_named_query() = data;
+    MessageToJsonString(element, &json);
+    elements_.push_back(json);
+    return json;
+  }
+
+  std::string AddDocumentMetadata(const ProtoBundledDocumentMetadata& data) {
+    std::string json;
+    ProtoBundleElement element;
+    *element.mutable_document_metadata() = data;
+    MessageToJsonString(element, &json);
+    elements_.push_back(json);
+    return json;
+  }
+
+  std::string AddDocument(const ProtoDocument& data) {
+    std::string json;
+    ProtoBundleElement element;
+    *element.mutable_document() = data;
+    MessageToJsonString(element, &json);
+    elements_.push_back(json);
+    return json;
+  }
+
+  std::string BuildBundle(const std::string& bundle_id,
+                          model::SnapshotVersion create_time,
+                          int32_t documents) {
+    std::string bundle;
+    for (const auto& element : elements_) {
+      auto bytes_length_string = std::to_string(element.size());
+      bundle.append(bytes_length_string);
+      bundle.append(element);
+    }
+
+    ProtoBundleMetadata metadata;
+    metadata.set_id(bundle_id);
+    metadata.set_version(1);
+    metadata.set_total_documents(documents);
+    metadata.mutable_create_time()->set_nanos(
+        create_time.timestamp().nanoseconds());
+    metadata.mutable_create_time()->set_seconds(
+        create_time.timestamp().seconds());
+    metadata.set_total_bytes(bundle.size());
+    ProtoBundleElement element;
+    *element.mutable_metadata() = metadata;
+
+    std::string metadata_str;
+    MessageToJsonString(element, &metadata_str);
+
+    return std::to_string(metadata_str.size()) + metadata_str + bundle;
+  }
+
+  ProtoNamedQuery LimitQuery() {
+    core::Query original = testutil::Query("bundles/docs/colls")
+                               .AddingFilter(testutil::Filter("foo", "==", 3))
+                               .AddingOrderBy(testutil::OrderBy("foo"))
+                               .WithLimitToFirst(1);
+    BundledQuery bundled_query(original.ToTarget(), core::LimitType::First);
+    NamedQuery named_query("limitQuery", bundled_query,
+                           testutil::Version(1000));
+    nanopb::ByteString bytes =
```

Comment:
Optional: consider using `auto` here and in similar statements below (`MakeByteString` strongly implies the type).